multiraft/multiraft.go

```diff
@@ -428,14 +436,18 @@ type state struct {
 	groups    map[proto.RangeID]*group
 	nodes     map[proto.RaftNodeID]*node
 	writeTask *writeTask
+	// buffer the event if s.Events channel is full to avoid the deadlock
+	// between s.Events channel and callbackChan.
+	pendingEvents []interface{}
```

Comment:
We now have two kinds of buffers for events: this slice and the buffer in the channel itself. I don't think it makes sense to keep both. Make the `Events` channel unbuffered and only write to it in the select loop(s) (removing the EventBufferSize setting). In `sendEvent`, always append to the slice and don't touch the channel.